File: model/field_and_rules.py
# -*- coding: utf-8 -*-
from random import randint, choice
from self_walkthrough import SelfWalkthrough

# В любом случае если в классе есть слово And, значит этот класс точно надо разбивать на два.
# Но я решил посмотреть почему именно так. Если свернуть все методы и свойства, то мы увидим,
# что у класса 4 публичных свойства и один магический метод.
# 3 из которых представляют поле (field), а одно представляет правила.
# Магический метод так же относится к полю.
# Единственное место где эти два класса пересекаются - это метод _initialize_self_walkthrough.
# Мне кажется это можно смело вынести в Game, а данный класс разделить.
# Причём таким образом: Field, RuleManager, Rule
class FieldAndRules(object):
    # м, поясняющая документация, отлично
    '''
    Этот класс генерирует финальное поле и логические условия игры.
    '''
    def __init__(self, comlexity):
        self._size = 6
        self._final_field = self._generate_final_field()
        self._rules = []
        self._defined_start_cells_count = comlexity
        self._defined_start_cells = []
        # вот generate_final_field возвращает значение и передаёт его
        # в поле класса, а данный метод всё делает внутри себя, в том числе и создаёт новое поле,
        # хотя общий смысл один: создать поле класса и заполнить его значением.
        # Нужно стараться быть последовательным и принимать одинаковые решения.
        # Мой совет: всегда инициализировать поля классов только в конструкторе, но я об этом уже
        # писал в других модулях.
        self._generate_start_rules()
        self._initialize_self_walkthrough()

    @property
    def field(self):
        return self._final_field

    @property
    def rules(self):
        return self._rules
 
    @property
    # Я бы заменил на start_cells
    def defined_start_cells(self):
        return self._defined_start_cells

    @property
    # пока не понятно зачем нужно отдельное свойство для получения количества стартовых ячеек
    # если есть свойство возвращающее список этих ячеек
    def defined_start_cells_count(self):
        return self._defined_start_cells_count

    def _generate_final_field(self):
        data = []
        for y in range(self._size):
            row = []
            count = self._size
            while count:
                # для лучшего понимания данную строку нужно
                # поместить в функцию generate_unique_number() и можно даже проверку на наличие в
                # строке поместить туда же. И в таком случае while count можно будет заменить на
                # for x in range(self._size) и таким образом будет понятней что первый цикл проходит
                # по строкам, а второй по столбцам.
                n = (y + 1) * 10 + randint(1, self._size)
                if n not in row:
                    row.append(n)
                    count -= 1
            data.append(row)
        return data

    def _generate_start_rules(self):
        # данный комментарий лучше перенести в документацию метода.
        self._min_nums = [] # 30 рандомных ячеек, по 5 из каждого ряда. Необходимый минимум для логических условий
        for row in range(6):  # а тут вы почему-то не используете self._size, когда измениться поле,
            # то можно получить неожиданный результат.
            random_inds = range(6)
            random_inds.remove(randint(0, 5))
            for column in random_inds:
                self._min_nums.append((row, column))

        for i in range(self._defined_start_cells_count): # создание уловий для определенных клеток при старте игры
            index_of_num = randint(0, len(self._min_nums) - 1)
            num = self._min_nums[index_of_num]
            rule = [self._final_field[num[0]][num[1]], 'define', num[1]]
            self._rules.append(rule)
            self._defined_start_cells.append(rule)
            self._min_nums.remove(num)

        while len(self._min_nums): # создание минимальных стартовых условий - если есть потребность
            # дать имя этому циклу, то это явный признак того, что он должен уйти в функцию. 
            for num in self._min_nums:
                if len(self._min_nums) != 1: # создание правила с 2 рандомными ячейками
                    self._min_nums.remove(num)
                    rand_num = choice(self._min_nums)
                    self._min_nums.remove(rand_num)
                    self._create_start_rule(num, rand_num)
                else: # создание правила с последней ячейкой (при наличии)
                    self._min_nums.remove(num)
                    self._create_start_rule(num, (randint(0, 5), randint(0, 5)))

    def _initialize_self_walkthrough(self):
        self._walkthrough_game = SelfWalkthrough(self._rules, self._size)
        self._walkthrough_game.try_to_win()
        while not self._walkthrough_game.is_won:
            self._run_update_rules()
        self._walkthrough_game.try_to_remove_rules(self._defined_start_cells_count)

    def _run_update_rules(self):
        '''
        Добавление уловий игры, пока игра точно не будет проходима.
        '''
        while True:
            r = self._create_rule(randint(1, 4)) # 1-4 тип условия игры. Это можно перевести в тип Enum
            # не понятно что такое 1 или 4, но если это будет соответствовать значению перечисления,
            # то с этим будет проще работать. А выбирать можно было бы через рандомную функцию choice
            if (r not in self._rules) and (reversed(r) not in self._rules):
                self._rules.append(r)
                self._walkthrough_game.updete_rules(self._rules)
                break

    def _create_start_rule(self, num1, num2): # num1 и num2 - кортеж вида (row, column)
        y1 = num1[0]
        x1 = num1[1]
        y2 = num2[0]
        x2 = num2[1]
        if x1 == x2: # в одом столбце
            rule_type = 1  # Возвращаясь к теме Enum, тут было бы понятней что это.
        elif abs(x1 - x2) == 1: # либо в соседних столбцах, либо лево-право
            rule_type = choice([2, 3, 3])
        elif abs(x1 - x2) == 2: # либо 3 в ряд, либо лево-право
            rule_type = choice([2, 4, 4])
            pass
        else: # лево-право
            rule_type = 2

        if rule_type == 1: # в одном столбце
            self._rules.append([self._final_field[y1][x1], '^', self._final_field[y2][x2]])
        elif rule_type == 2: # лево-право
            if x1 < x2:
                self._rules.append([self._final_field[y1][x1], '...', self._final_field[y2][x2]])
            elif x1 > x2:
                self._rules.append([self._final_field[y2][x2], '...', self._final_field[y1][x1]])
        elif rule_type == 3: # в соседних столбцах
            self._rules.append([self._final_field[y1][x1], '<->', self._final_field[y2][x2]])
        elif rule_type == 4: # 3 в ряд
            x_mid = (x1 + x2) / 2
            y_mid = randint(0, 5)
            if (y_mid, x_mid) in self._min_nums:
                self._min_nums.remove((y_mid, x_mid))
            self._rules.append([self._final_field[y1][x1], self._final_field[y_mid][x_mid], self._final_field[y2][x2]])

    def _create_rule(self, rule_type): # создание рандомного правила
        y = randint(0, 5)
        x = randint(0, 5)
        if rule_type == 1: # в одном столбце
            y2 = y
            while y2 == y:
                y2 = randint(0, 5)
            return [self._final_field[y][x], '^', self._final_field[y2][x]]
        elif rule_type == 2: # порядок
            x1 = randint(0, 4)
            x2 = randint(x1 + 1, 5)
            y2 = randint(0, 5)
            return [self._final_field[y][x1], '...', self._final_field[y2][x2]]
        elif rule_type == 3: # рядом
            if x == 0:
                x2 = 1
            elif x == 5:
                x2 = 4
            else:
                x2 = x + choice([-1, 1])
            y2 = randint(0, 5)
            return [self._final_field[y][x], '<->', self._final_field[y2][x2]]
        elif rule_type == 4: # 3 в ряд
            if x < 2:
                dx = 1
            elif x > 3:
                dx = -1
            else:
                dx = choice([-1, 1])
            x2 = x + dx
            x3 = x2 + dx
            y2 = randint(0, 5)
            y3 = randint(0, 5)
            return [self._final_field[y][x], self._final_field[y2][x2], self._final_field[y3][x3]]

    def __getitem__(self, index):
        return self._final_field[index]

File: phaser/src/model/fieldAndRules.ts
import { Rule } from './types';
import { Rng } from './rng';
import { SelfWalkthrough } from './selfWalkthrough';

// Pre-revealed starter cells per (size, difficulty). Easy hands a big head
// start; Hard reveals nothing. Mirror of presenter/game.py _COMPLEXITY.
export const COMPLEXITY: Record<number, [number, number, number]> = {
  4: [9, 4, 0],
  5: [14, 7, 0],
  6: [20, 10, 0],
};

const MAX_RULE_ADDS = 600;
const MAX_RULE_ATTEMPTS = 400;

/**
 * Faithful port of model/field_and_rules.py.
 *
 * Generates the solved board (`field`), a set of clues (`rules`) and the
 * revealed starter cells (`definedStartCells`). The rule set is grown with
 * random clues until {@link SelfWalkthrough} can fully deduce the board, then
 * trimmed of redundant clues — so every generated puzzle is solvable without
 * guessing.
 */
export class FieldAndRules {
  readonly seed: number;
  readonly size: number;
  readonly field: number[][];
  readonly rules: Rule[] = [];
  readonly definedStartCells: Rule[] = [];
  readonly definedStartCellsCount: number;

  private readonly rng: Rng;
  private minNums: Array<[number, number]> = [];
  private solver!: SelfWalkthrough;

  constructor(complexity: number, size: number, seed: number) {
    this.seed = seed;
    this.size = size;
    this.rng = new Rng(seed);
    this.field = this.generateFinalField();
    this.definedStartCellsCount = complexity;
    this.generateStartRules();
    this.initSolver();
  }

  /** The solved board doubles as the answer grid for the player's cascade. */
  get solution(): number[][] {
    return this.field;
  }

  /** Clues shown to the player (the revealed starter cells are excluded — the
   *  board pre-solves those instead). */
  get displayableRules(): Rule[] {
    return this.rules.slice(this.definedStartCellsCount);
  }

  private generateFinalField(): number[][] {
    const data: number[][] = [];
    for (let y = 0; y < this.size; y++) {
      const row: number[] = [];
      let count = this.size;
      while (count) {
        const n = (y + 1) * 10 + this.rng.randint(1, this.size);
        if (!row.includes(n)) {
          row.push(n);
          count -= 1;
        }
      }
      data.push(row);
    }
    return data;
  }

  private generateStartRules(): void {
    // N*(N-1) cells, N-1 per row — the minimum pool the clues are built from.
    this.minNums = [];
    for (let row = 0; row < this.size; row++) {
      const inds: number[] = [];
      for (let i = 0; i < this.size; i++) inds.push(i);
      const drop = this.rng.randint(0, this.size - 1);
      inds.splice(inds.indexOf(drop), 1);
      for (const column of inds) this.minNums.push([row, column]);
    }

    // revealed starter cells
    for (let i = 0; i < this.definedStartCellsCount; i++) {
      const idx = this.rng.randint(0, this.minNums.length - 1);
      const num = this.minNums[idx];
      const rule: Rule = [this.field[num[0]][num[1]], 'define', num[1]];
      this.rules.push(rule);
      this.definedStartCells.push(rule);
      this.minNums.splice(idx, 1);
    }

    // pair the rest into minimal starting clues
    while (this.minNums.length) {
      if (this.minNums.length === 1) {
        const num = this.minNums.shift()!;
        this.createStartRule(num, [
          this.rng.randint(0, this.size - 1),
          this.rng.randint(0, this.size - 1),
        ]);
      } else {
        const num = this.minNums.shift()!;
        const ri = this.rng.randint(0, this.minNums.length - 1);
        const rand = this.minNums.splice(ri, 1)[0];
        this.createStartRule(num, rand);
      }
    }
  }

  private initSolver(): void {
    this.solver = new SelfWalkthrough(this.rules, this.size);
    this.solver.tryToWin();
    let adds = 0;
    while (!this.solver.isWon) {
      if (adds >= MAX_RULE_ADDS) {
        throw new Error(
          `FieldAndRules: gave up generating rules for seed ${this.seed} (size=${this.size})`,
        );
      }
      this.runUpdateRules();
      adds += 1;
    }
    this.solver.tryToRemoveRules(this.definedStartCellsCount);
  }

  private runUpdateRules(): void {
    for (let i = 0; i < MAX_RULE_ATTEMPTS; i++) {
      const r = this.createRule(this.rng.randint(1, 4));
      if (!this.ruleExists(r)) {
        this.rules.push(r);
        this.solver.updateRules(this.rules);
        return;
      }
    }
    throw new Error(
      `FieldAndRules: ${MAX_RULE_ATTEMPTS} rule-create attempts produced only duplicates (seed=${this.seed}, size=${this.size})`,
    );
  }

  private ruleExists(r: Rule): boolean {
    return this.rules.some(
      (e) => e.length === r.length && e.every((v, i) => v === r[i]),
    );
  }

  // num1/num2 are [row, column]
  private createStartRule(num1: [number, number], num2: [number, number]): void {
    const [y1, x1] = num1;
    const [y2, x2] = num2;
    let ruleType: number;
    if (x1 === x2) ruleType = 1;
    else if (Math.abs(x1 - x2) === 1) ruleType = this.rng.choice([2, 3, 3]);
    else if (Math.abs(x1 - x2) === 2) ruleType = this.rng.choice([2, 4, 4]);
    else ruleType = 2;

    if (ruleType === 1) {
      this.rules.push([this.field[y1][x1], '^', this.field[y2][x2]]);
    } else if (ruleType === 2) {
      if (x1 < x2) this.rules.push([this.field[y1][x1], '...', this.field[y2][x2]]);
      else if (x1 > x2) this.rules.push([this.field[y2][x2], '...', this.field[y1][x1]]);
    } else if (ruleType === 3) {
      this.rules.push([this.field[y1][x1], '<->', this.field[y2][x2]]);
    } else if (ruleType === 4) {
      const xMid = Math.floor((x1 + x2) / 2);
      const yMid = this.rng.randint(0, this.size - 1);
      const k = this.minNums.findIndex((p) => p[0] === yMid && p[1] === xMid);
      if (k >= 0) this.minNums.splice(k, 1);
      this.rules.push([this.field[y1][x1], this.field[yMid][xMid], this.field[y2][x2]]);
    }
  }

  private createRule(ruleType: number): Rule {
    const last = this.size - 1;
    const y = this.rng.randint(0, last);
    const x = this.rng.randint(0, last);
    if (ruleType === 1) {
      let y2 = y;
      while (y2 === y) y2 = this.rng.randint(0, last);
      return [this.field[y][x], '^', this.field[y2][x]];
    }
    if (ruleType === 2) {
      const x1 = this.rng.randint(0, last - 1);
      const x2 = this.rng.randint(x1 + 1, last);
      const y2 = this.rng.randint(0, last);
      return [this.field[y][x1], '...', this.field[y2][x2]];
    }
    if (ruleType === 3) {
      let x2: number;
      if (x === 0) x2 = 1;
      else if (x === last) x2 = last - 1;
      else x2 = x + this.rng.choice([-1, 1]);
      const y2 = this.rng.randint(0, last);
      return [this.field[y][x], '<->', this.field[y2][x2]];
    }
    // ruleType === 4 : three in a row
    let dx: number;
    if (x <= 1) dx = 1;
    else if (x >= this.size - 2) dx = -1;
    else dx = this.rng.choice([-1, 1]);
    const x2 = x + dx;
    const x3 = x2 + dx;
    const y2 = this.rng.randint(0, last);
    const y3 = this.rng.randint(0, last);
    return [this.field[y][x], this.field[y2][x2], this.field[y3][x3]];
  }
}

/** Build a puzzle, re-rolling the seed if the generator's safety caps trip. */
export function generatePuzzle(
  size: number,
  difficulty: number,
  seed?: number,
): { model: FieldAndRules; seed: number } {
  let s = seed ?? 1 + Math.floor(Math.random() * (2 ** 31 - 1));
  const complexity = COMPLEXITY[size][difficulty];
  let lastErr: unknown;
  for (let attempt = 0; attempt < 10; attempt++) {
    try {
      return { model: new FieldAndRules(complexity, size, s), seed: s };
    } catch (e) {
      lastErr = e;
      s = ((s * 1103515245 + 12345) & 0x7fffffff) || 1;
    }
  }
  throw lastErr instanceof Error ? lastErr : new Error('puzzle generation failed');
}
